ls1intum/Artemis
# `General`: Allow to publish example solutions for modeling, text and file upload exercises on a specific date

### Checklist
#### General


- [x] I tested **all** changes and their related features with **all** (student, instructor) corresponding user types ~on a test server~ locally.
#### Server
- [x] (No new REST calls) I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
#### Client
- [x] I added multiple **unit** tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] (No new routes) I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).

### Motivation and Context


This PR solves https://github.com/ls1intum/Artemis/issues/16 for the modeling, text and file upload exercises. It allows example solutions (previously called "sample solution") to be published to the students after a predefined date.

This functionality will be added to the programming exercises in a follow-up PR.

### Description

Note: Contains a database migration.

This PR adds `example_solution_publication_date` to `Exercise` model and allows instructors to set a date to make the example solution that is already present in the database available to the students for modeling, text and file upload exercises. Currently presenting example solutions is explicitly disabled for exam exercises, it is planned to extend this feature to include them in a follow-up PR. Also, sampleSolution (and similar) fields, attributes, translation strings, etc. are renamed to exampleSolution (and similar) for Modeling, Text and File Upload Exercises because "Example solution" term is preferred over "Sample solution".

### Steps for Testing

Contains a migration so please deploy only to TS1 or TS3 (new).
Prerequisites:
- 1 Instructor
- 1 Students
- 1 Modeling Exercise with a Example Solution
- 1 Text Exercise with a Example Solution
- 1 File Upload Exercise with a Example Solution

Instructor:

1. Log in to Artemis
2. Navigate to Course Administration
3. Go to the Edit page of the prerequisite Modeling Exercise.
4. Ensure current `Example Solution Publication Date` is empty or after current time.

Student:

5. Log in to Artemis
6. Navigate to the prerequisite Modeling Exercise's detail page.
7. Ensure example solution is not visible.

Instructor:

8. Change `Example Solution Publication Date` of the exercise to a time before current time.

Student:

9. Ensure example solution is visible.
10. Ensure the term "Example solution" is used instead "Sample solution".

11. Repeat steps 1-9 with `Modeling Exercise` replaced with `Text Exercise` and `File Upload Exercise`.

Instructor:

12. Import an existing `Modeling Exercise` with `Example Solution Publication Date` set to a non-null date.
13. Ensure new exercise has `Example Solution Publication Date` set to null.

14. Repeat steps 12-13 with `Modeling Exercise` replaced with `Text Exercise`.

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage





File: Branches % | Lines %

BaseExercise.java: 86% | 99%
FileUploadExercise.java: 100% | 100%
TextExercise.java: 100% | 100%
ModelingExercise.java: 100% | 89%
course-exercise-details.component.ts : 59.23% | 79.6%
modeling-exercise-update.component.ts: 81.92% | 92.07%
### Screenshots
#### Course Management View (For Modeling Exercise)



![image](https://user-images.githubusercontent.com/18427209/147265557-97ac7d8c-c555-463d-9e85-ac5027f69377.png)

![image](https://user-images.githubusercontent.com/18427209/147265369-619c112c-4451-4941-afde-116f28d1fe3a.png)

#### Course Exercise Detail
##### Modeling Exercise

![image](https://user-images.githubusercontent.com/18427209/142378655-156a086c-9ec7-4712-8315-7e6a5982b982.png)

![image](https://user-images.githubusercontent.com/18427209/142378710-c225893a-7a8b-4ea7-92f9-88e2b732754e.png)


##### Text Exercise
![image](https://user-images.githubusercontent.com/18427209/147660866-a3ecd768-90a6-46b7-beb4-5ac0dc569981.png)

##### File Upload Exercise
![image](https://user-images.githubusercontent.com/18427209/147660924-ecc1a507-721b-45ed-b262-b368082fdefd.png)
